mmdet/datasets/pipelines/copy_paste_aug.py

```diff
@@ -0,0 +1,162 @@
+import cv2
+import random
+import numpy as np
+from copy import deepcopy
+
+from .auto_augment import Rotate
+from .transforms import Resize, RandomFlip
+from ..builder import PIPELINES
+
+
+@PIPELINES.register_module()
+class CopyPaste:
+    def __init__(
+        self,
+        copy_all=True,
+        horizontal_flip_ratio=0.5,
+        max_rotate_angle=30,
+        rotate_level=2,
+        rotate_prob=0.5
+    ):
+        self.copy_all = copy_all
+        self.resize = None
+        self.flip = RandomFlip(flip_ratio=horizontal_flip_ratio)
+        if max_rotate_angle > 0:
+            self.rotate = Rotate(level=rotate_level, prob=rotate_prob, max_rotate_angle=max_rotate_angle)
+        else:
+            self.rotate = None
+        self.colors = [(np.random.rand(3) * 255).astype(np.uint8) for _ in range(20)]
+
+    def image_copy_paste(self, results, composed_mask):
+        img_dtype = results['img'].dtype
+        composed_mask = composed_mask[..., None]
+        results['img'] = results['copy_paste']['img'] * composed_mask + results['img'] * (1 - composed_mask)
+        results['img'] = results['img'].astype(img_dtype)
+
+    def masks_copy_paste(self, results, paste_objects, composed_mask):
+        results['gt_masks'].masks = np.array([
+            np.where(composed_mask == 1, 0, mask).astype(np.uint8) for mask in results['gt_masks'].masks
+        ])
+        results['gt_masks'].masks = np.concatenate(
+            (results['gt_masks'].masks, results['copy_paste']['gt_masks'].masks[paste_objects]), axis=0)
+        results['gt_bboxes'] = np.concatenate(
+            (results['gt_bboxes'], results['copy_paste']['gt_bboxes'][paste_objects]), axis=0)
+        self.extract_bboxes(results)
+
+    @staticmethod
+    def concatenate_labels(results, paste_objects=None):
+        results['gt_labels'] = np.concatenate(
+            (results['gt_labels'], results['copy_paste']['gt_labels'][paste_objects]), axis=0)
+        results['gt_bboxes_ignore'] = np.concatenate(
+            (results['gt_bboxes_ignore'], results['copy_paste']['gt_bboxes_ignore']), axis=0)
```

Comment:
Are ignored boxes/masks copied from 'copy_paste' image to the target one? Should `paste_objects` be taken into account somehow?

Reply:
Sounds reasonable. But I am not sure that this will give something. So, added filter for objects that are inside ignore regions as option.